### PR TITLE
CORE: improve logging

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Improve logging in different components of UCC

## Why ?
Currently there is no unified rules between TLs, CLs, CORE about what to print and how to print.

## How ?
1. Make LOG_LEVE=info less verbose, print only essential info about UCC runtime including
1.1. UCC version
1.2. UCC runtime path (path to the libucc.so)
1.3. global config path if used or n/a otherwise
1.4. team create log with team id and team size
1.5. score map i.e. what collectives available, what memory types available, what TLs/CLs are used
1.6. team destroy log
2. Added UCC_COLL_TRACE (default is no)
2.1. Leader rank prints collective info on each ucc_collective_init if UCC log level is info
2.2. All ranks from team print collective info on ucc_collective_init if UCC log level is debug or above
2.3. For LOG_LEVEL=info collective info includes, src and dst buffer address, count, datatype, memory type if available; inplace flag if set; root for rooted collectives; operation for reduction collectives; chosen TL/CL; team id. For V collectives (allgatherv, alltoallv, scatterv ...) count is sum of corresponding values in counts buffer.
2.4. For LOG_LEVEL=debug collective info output additionally includes team rank number, context rank number, collective sequence number, pointer to request

All other prints moved from info log level to debug log level.
<details>
<summary>UCC_LOG_LEVEL=info example</summary>
<pre>
ucc_constructor.c:183  UCC  INFO  version: 1.2.0, loaded from: /home/ucc/build-debug/install/lib/libucc.so.1, cfg file: n/a
ucc_constructor.c:183  UCC  INFO  version: 1.2.0, loaded from: /home/ucc/build-debug/install/lib/libucc.so.1, cfg file: n/a
       ucc_team.c:471  UCC  INFO  ===== COLL_SCORE_MAP (team_id 32768, size 2) =====
ucc_coll_score_map.c:201  UCC  INFO  Allgather:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Allgatherv:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Allreduce:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..4095}:TL_UCP:10 {4K..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..4095}:TL_UCP:10 {4K..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..4095}:TL_UCP:10 {4K..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Alltoall:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Alltoallv:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Barrier:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Bcast:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..32767}:TL_UCP:10 {32K..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..32767}:TL_UCP:10 {32K..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..32767}:TL_UCP:10 {32K..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Fanin:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Fanout:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Gather:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Gatherv:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce_scatter:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce_scatterv:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Scatterv:
ucc_coll_score_map.c:201  UCC  INFO     Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     Cuda: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO     CudaManaged: {0..inf}:TL_UCP:10
       ucc_team.c:474  UCC  INFO  ================================================

 OSU MPI All-to-All Personalized Exchange Latency Test v5.8
 Size       Avg Latency(us)
....
ucc_team.c:528  UCC  INFO  team destroyed, team_id 32768
</pre>
</details>

<details>
<summary>UCC_COLL_TRACE=y example</summary>
<pre>
ucc_coll.c:257  UCC  INFO  coll_init: Barrier; TL_UCP, team_id 32768
ucc_coll.c:257  UCC  INFO  coll_init: Alltoall: src={0x7f7f2d789000, 2097152, int8, Host}, dst={0x7f7f2d587000, 2097152, int8, Host}; TL_UCP, team_id 32768
ucc_coll.c:257  UCC  INFO  coll_init: Barrier; TL_UCP, team_id 32768
ucc_coll.c:257  UCC  INFO  coll_init: Reduce min root 0: src={0x7ffd97e4ebd0, 1, float64, Host}, dst={0x7ffd97e4ebe8, 1, float64, Host}; TL_UCP, team_id 32768
ucc_coll.c:257  UCC  INFO  coll_init: Reduce max root 0: src={0x7ffd97e4ebd0, 1, float64, Host}, dst={0x7ffd97e4ebe0, 1, float64, Host}; TL_UCP, team_id 32768
ucc_coll.c:257  UCC  INFO  coll_init: Reduce sum root 0: src={0x7ffd97e4ebd0, 1, float64, Host}, dst={0x7ffd97e4ebd8, 1, float64, Host}; TL_UCP, team_id 32768
</pre>
</details>

### CODE REVIEW

**Comment by Sergei-Lebedev** [comment]
ok to test

**Comment by Sergei-Lebedev** [comment]
@jirikraus please take a look

**Comment by Sergei-Lebedev** [comment]
bot:retest

**Comment by shimmybalsam** [comment]
General: in the print you show from UCC_LOG_LEVEL=info example there is no CL information, only TL score map. perhaps chosen CL info should be added as well.

File: src/core/ucc_global_opts.c
Comment (shimmybalsam) [inline src/core/ucc_global_opts.c:34]: a bit confusing since you're explaining what would happen in case of "y" but defining as "n"

**Review by shimmybalsam — APPROVED** [review]

File: src/core/ucc_global_opts.c
Comment (shimmybalsam) [inline src/core/ucc_global_opts.c:34]: a bit confusing since you're explaining what would happen in case of "y" but defining as "n"

**Review by shimmybalsam — CHANGES_REQUESTED** [review]

**Comment by jirikraus** [comment]
> 

Thanks @Sergei-Lebedev looking at the example output this looks great. I can't comment on the source change but assume this is not requested. As @shimmybalsam points out I think it makes sense to also cover the choosen CL.

**Review by samnordmann — APPROVED** [review]

File: src/utils/ucc_coll_utils.c
Comment (vspetrov) [inline src/utils/ucc_coll_utils.c:390]: maybe move thise entire blob into separate util function:
```
static inline void ucc_coll_args_get_info(const ucc_coll_args_t *args, ucc_coll_buffer_info_t *s_info, 
ucc_coll_buffer_info_t *d_info) {
...
}
```
it can set s_info.buffer or r_info.buffer to NULL that would indicate the same as has_src/has_dst flags.

**Review by vspetrov — COMMENTED** [review]
The change is awesome! I have just minor comments. 
Why not keeping coll seq_num on the trace with INFO level. the line will be almost same length? It gives a lot of info on how heavily the team is used (if seq_num is large or not), etc.

Looks like we don't have trace for triggered post, lets add.

A couple of next steps we could as a follow up:
* Add UCC_TL_COLL_TRACE 
basically same tracing but at TL level. This var should of course NOT inherit from UCC_COLL_TRACE. It would be convenient to trace in detail actual algorithms start/finalize at TL level.

* With existing tracing, what else do we need to add to logging to potentially "replay" the trace ? This could be a very handy tool for perf analysis

File: src/core/ucc_global_opts.c
Comment (Sergei-Lebedev) [reply src/core/ucc_global_opts.c:34]: why is it confusing? by default coll trace is disabled because it produces a lot of output. If you want to see it just set UCC_COLL_TRACE=y.

**Comment by Sergei-Lebedev** [comment]
> Why not keeping coll seq_num on the trace with INFO level. the line will be almost same length? It gives a lot of info on how heavily the team is used (if seq_num is large or not), etc.

Sure, no problem, will add.

> Looks like we don't have trace for triggered post, lets add.

We print this info in coll_init so we will have it for triggered post as well. But maybe move it to collective post and triggered post instead? I'm not sure which option is better. 


> * Add UCC_TL_COLL_TRACE
>   basically same tracing but at TL level. This var should of course NOT inherit from UCC_COLL_TRACE. It would be convenient to trace in detail actual algorithms start/finalize at TL level.

Agree, but we have much more option where to add in TL. For example in every init, post, frag start ... 

> * With existing tracing, what else do we need to add to logging to potentially "replay" the trace ? This could be a very handy tool for perf analysis

Yes, i wanted to start with some script that can collect different statistic, but replay is also an option.

File: src/utils/ucc_coll_utils.c
Comment (Sergei-Lebedev) [reply src/utils/ucc_coll_utils.c:390]: s_info.buffer == NULL is actually valid buffer, for example we did alltoallv with 64bit displacement in pytorch

**Comment by jirikraus** [comment]
> 

I think that would address my point. Can you provide an example output to make your proposal more concrete?

**Comment by jirikraus** [comment]
> > Thanks @Sergei-Lebedev looking at the example output this looks great. I can't comment on the source change but assume this is not requested. As @shimmybalsam points out I think it makes sense to also cover the choosen CL.
> 
> It prints CL HIER if it's used and some TL if CL BASIC is used. The logic is that we print top level task info which is TL task for CL BASIC and CL task for CL HIER.

I think that would address my point. However I am lacking background to figure how this might look like in practice. Can you provide an example how that would look? The example you provide above only cover TL_UCP not the CL used.

**Comment by Sergei-Lebedev** [comment]
> > > Thanks @Sergei-Lebedev looking at the example output this looks great. I can't comment on the source change but assume this is not requested. As @shimmybalsam points out I think it makes sense to also cover the choosen CL.
> > 
> > 
> > It prints CL HIER if it's used and some TL if CL BASIC is used. The logic is that we print top level task info which is TL task for CL BASIC and CL task for CL HIER.
> 
> I think that would address my point. However I am lacking background to figure how this might look like in practice. Can you provide an example how that would look? The example you provide above only cover TL_UCP not the CL used.

sure, it's similar:
```
ucc_coll.c:257  UCC  INFO  coll_init: Barrier; CL_HIER, team_id 1
ucc_coll.c:257  UCC  INFO  coll_init: Allreduce min: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8228, 1, float64, Host}; CL_HIER, team_id 1
ucc_coll.c:257  UCC  INFO  coll_init: Allreduce max: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8220, 1, float64, Host}; CL_HIER, team_id 1
ucc_coll.c:257  UCC  INFO  coll_init: Allreduce sum: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8230, 1, float64, Host}; CL_HIER, team_id 1
```

**Comment by shimmybalsam** [comment]
> > > > Thanks @Sergei-Lebedev looking at the example output this looks great. I can't comment on the source change but assume this is not requested. As @shimmybalsam points out I think it makes sense to also cover the choosen CL.
> > > 
> > > 
> > > It prints CL HIER if it's used and some TL if CL BASIC is used. The logic is that we print top level task info which is TL task for CL BASIC and CL task for CL HIER.
> > 
> > 
> > I think that would address my point. However I am lacking background to figure how this might look like in practice. Can you provide an example how that would look? The example you provide above only cover TL_UCP not the CL used.
> 
> sure, it's similar:
> 
> ```
> ucc_coll.c:257  UCC  INFO  coll_init: Barrier; CL_HIER, team_id 1
> ucc_coll.c:257  UCC  INFO  coll_init: Allreduce min: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8228, 1, float64, Host}; CL_HIER, team_id 1
> ucc_coll.c:257  UCC  INFO  coll_init: Allreduce max: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8220, 1, float64, Host}; CL_HIER, team_id 1
> ucc_coll.c:257  UCC  INFO  coll_init: Allreduce sum: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8230, 1, float64, Host}; CL_HIER, team_id 1
> ```

Looks good but I still think that also for CL BASIC we should output that we are using CL_BASIC and then which TLs

**Comment by jirikraus** [comment]
> > > > > Thanks @Sergei-Lebedev looking at the example output this looks great. I can't comment on the source change but assume this is not requested. As @shimmybalsam points out I think it makes sense to also cover the choosen CL.
> > > > 
> > > > 
> > > > It prints CL HIER if it's used and some TL if CL BASIC is used. The logic is that we print top level task info which is TL task for CL BASIC and CL task for CL HIER.
> > > 
> > > 
> > > I think that would address my point. However I am lacking background to figure how this might look like in practice. Can you provide an example how that would look? The example you provide above only cover TL_UCP not the CL used.
> > 
> > 
> > sure, it's similar:
> > ```
> > ucc_coll.c:257  UCC  INFO  coll_init: Barrier; CL_HIER, team_id 1
> > ucc_coll.c:257  UCC  INFO  coll_init: Allreduce min: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8228, 1, float64, Host}; CL_HIER, team_id 1
> > ucc_coll.c:257  UCC  INFO  coll_init: Allreduce max: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8220, 1, float64, Host}; CL_HIER, team_id 1
> > ucc_coll.c:257  UCC  INFO  coll_init: Allreduce sum: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8230, 1, float64, Host}; CL_HIER, team_id 1
> > ```
> 
> Looks good but I still think that also for CL BASIC we should output that we are using CL_BASIC and then which TLs

And as @vspetrov suggested add UCC_TL_COLL_TRACE to provide additional details. To make sure I understand this right when CL hier is used we won't see the TLs it leverages? Could we show that without getting too verbose, E.g.

```
ucc_coll.c:257  UCC  INFO  coll_init: Barrier; CL_HIER, team_id 1
ucc_coll.c:257  UCC  INFO  coll_init: Allreduce min: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8228, 1, float64, Host}; CL_HIER ( comma seperated list of TLs levearge), team_id 1
ucc_coll.c:257  UCC  INFO  coll_init: Allreduce max: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8220, 1, float64, Host}; CL_HIER ( comma seperated list of TLs levearge), team_id 1
ucc_coll.c:257  UCC  INFO  coll_init: Allreduce sum: src={0x7ffd443b8238, 1, float64, Host}, dst={0x7ffd443b8230, 1, float64, Host}; CL_HIER ( comma seperated list of TLs levearge, e.g. (TL_UCP, TL_SHARP)) , team_id 1
```

and something similar for CL Basic?

File: src/core/ucc_global_opts.c
Comment (shimmybalsam) [reply src/core/ucc_global_opts.c:34]: In my opinion it would be more clear like this:
"If enabled and UCC logging level is INFO or higher UCC will print info message per collective.\n"

File: src/utils/ucc_coll_utils.c
Comment (vspetrov) [reply src/utils/ucc_coll_utils.c:390]: you are right. still worth moving to separate fn imho, even with 2 more flags. current function is very large. just from readability perspective.

**Comment by vspetrov** [comment]
> We print this info in coll_init so we will have it for triggered post as well. But maybe move it to collective post and triggered post instead? I'm not sure which option is better.
> 

right, i'm starting to forget stuff... i think, we should keep as is - print detailed info on collective_init and on post/finalize we have a verbose messages that will be printed under DEBUG. BTW, in post/finalize we print pointer of the request which makes it easy to match initialized request with the posted/finalized. I saw you no longer print the request pointer in coll_init. Maybe lets add it under DEBUG.

**Review by shimmybalsam — APPROVED** [review]

**Comment by jirikraus** [comment]
> @jirikraus @vspetrov @shimmybalsam I slightly changed the log for collective trace. Now UCC_COLL_TRACE is enum instead of binary value, which allows to finer control over verbosity level. Coll trace doesn't depend on main LOG_LEVEL anymore, examples

Thanks this looks great. @Sergei-Lebedev when is this expected to land in a UCC release? I wonder how we should track adding a description of this to the [user_guide.md](https://github.com/jirikraus/ucc/blob/ucc_userguide_proposal/docs/user_guide.md) discussed in PR #720 ?

**Comment by Sergei-Lebedev** [comment]
> > @jirikraus @vspetrov @shimmybalsam I slightly changed the log for collective trace. Now UCC_COLL_TRACE is enum instead of binary value, which allows to finer control over verbosity level. Coll trace doesn't depend on main LOG_LEVEL anymore, examples
> 
> Thanks this looks great. @Sergei-Lebedev when is this expected to land in a UCC release? I wonder how we should track adding a description of this to the [user_guide.md](https://github.com/jirikraus/ucc/blob/ucc_userguide_proposal/docs/user_guide.md) discussed in PR #720 ?

If everybody agrees on logging format, then this PR will be merged pretty soon. After it's merged you can update User Guide accordingly and I don't think it depends on release date. But answering your question we are going to release new UCC version in April.

**Comment by Sergei-Lebedev** [comment]
@vspetrov @bureddy kind remainder, pls approve if you are ok with the changes

**Review by vspetrov — APPROVED** [review]

**Review by manjugv — APPROVED** [review]

**Pull request merged**
